### PR TITLE
INT-4417: Add @EndpointId

### PR DESCRIPTION
**Pull request by garyrussell**
JIRA: https://jira.spring.io/browse/INT-4417

Provide more flexibility with bean naming when using EIP Annotations with
Java configuration.

### CODE REVIEW

File: spring-integration-core/src/main/java/org/springframework/integration/config/annotation/MessagingAnnotationPostProcessor.java [hunk 1]
```diff
@@ -255,16 +252,24 @@ protected boolean recursiveFindAnnotation(Class<? extends Annotation> annotation
...
+			name = baseName;
+			int count = 1;
+			while (this.beanFactory.containsBean(name)) {
+				name = baseName + "#" + (++count);
+			}
 		}
 		return name;
 	}
 
+	public static String endpointIdValue(Method method) {
```
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/config/annotation/MessagingAnnotationPostProcessor.java]: Can we move this utility into the `MessagingAnnotationUtils` ?

File: src/reference/asciidoc/configuration.adoc [hunk 2]
```diff
@@ -612,12 +613,13 @@ This works like there is no Messaging Annotation on the `@Bean` method.
 * The `AbstractEndpoint` bean name is generated with the pattern:
 `[configurationComponentName].[methodName].[decapitalizedAnnotationClassShortName]`.
 For example the endpoint (`SourcePollingChannelAdapter`) for the `consoleSource()` definition above gets a bean name like:
-`myFlowConfiguration.consoleSource.inboundChannelAdapter`.
+`myFlowConfiguration.consoleSource.inboundChannelAdapter`. Also see <<endpoint-bean-names>>.
```
Comment (artembilan) [inline src/reference/asciidoc/configuration.adoc]: New sentence from a new line.

**Review by artembilan — CHANGES_REQUESTED** [review]
Nothing critical though

**Comment by artembilan** [comment]
Ping

File: spring-integration-core/src/main/java/org/springframework/integration/annotation/EndpointId.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+ * conventions in the framework. Similarly, for a message source, use
+ * {@code @Bean("bar.source"} and {@code @EndpointId("bar")}.
+ *
+ * @author Gary Russell
+ *
+ * @since 5.0.4
+ */
+@Target({ElementType.METHOD, ElementType.ANNOTATION_TYPE})
+@Retention(RetentionPolicy.RUNTIME)
+@Inherited
```
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/annotation/EndpointId.java]: Not sure that `@Inherited` is OK here. 
If you override method and then register it in the application context with some Messaging Annotation, you definitely have to have its own distinct endpoint id. Therefore `@EndpointId` is required on the overridden method.

WDYT?

**Review by artembilan — CHANGES_REQUESTED** [review]
While it is a good conventional addition I still think there is a use-case when we can have several Messaging Annotations on the same method. Right, it sounds awkward and might not have a logical reason, but when we add this `@EndpointId` to such a definition we end up with the when it will be overridden with the unique bean name based on the index.

Saying that we might need to document the order how those annotations are parsed:
```
this.postProcessors.put(Filter.class, );
this.postProcessors.put(Router.class, );
this.postProcessors.put(Transformer.class, );
this.postProcessors.put(ServiceActivator.class, );
this.postProcessors.put(Splitter.class, );
this.postProcessors.put(Aggregator.class, );
```

Or just disallow to configure `@EndpointId` on several Messaging Annotations.

**Comment by artembilan** [comment]
See `MessagingAnnotationsWithBeanAnnotationTests.ContextConfiguration.skippedMessageHandler()` bean definition for more info.

File: spring-integration-core/src/main/java/org/springframework/integration/annotation/EndpointId.java (hunk 3 above)
Comment (garyrussell) [reply spring-integration-core/src/main/java/org/springframework/integration/annotation/EndpointId.java]: Looks like this is wrong on several of the EIP annotations...
```
 * <p>Note that this meta-annotation type has no effect if the annotated
 * type is used to annotate anything other than a class.  Note also
 * that this meta-annotation only causes annotations to be inherited
 * from superclasses; annotations on implemented interfaces have no
 * effect.
```

Comment (artembilan) [reply spring-integration-core/src/main/java/org/springframework/integration/annotation/EndpointId.java]: Hm. So, sounds like fixing that won't cause any breaking changes.
Isn't it ?

**Review by artembilan — APPROVED** [review]
OK. That's cool.

Will we remove `@Inherited` from other EIP method-based annotations in `5.1` ?

**Comment by garyrussell** [comment]
Yes; I think we should wait for 5.1.

**Comment by artembilan** [comment]
>  I think we should wait for 5.1.

Good. I'm raising JIRA then...

**Comment by artembilan** [comment]
For better traceability the JIRA about `@Inherited` on method-based annotations: https://jira.spring.io/browse/INT-4424

**Pull request merged**